File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/nonterminal/upperscopes/MethodNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.upperscopes;

import org.vanautrui.languages.compiler.parsing.IASTNode;
import org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.DeclaredArgumentNode;
import org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.statements.StatementNode;
import org.vanautrui.languages.compiler.parsing.astnodes.typenodes.TypeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public final class MethodNode implements IASTNode {

	public final boolean isPublic = true;

	public final boolean has_side_effects;

	public final TypeNode returnType;

	public final String methodName;

	public final List<DeclaredArgumentNode> arguments = new ArrayList<>();

	public final List<StatementNode> statements = new ArrayList<>();

	public MethodNode(final TypeNode returnType, final String methodName) {
		this.returnType = returnType;
		this.methodName = methodName;
		this.has_side_effects = false;
	}

	public MethodNode(final TypeNode returnType, final String methodName, final List<StatementNode> statements) {
		this.returnType = returnType;
		this.methodName = methodName;
		this.has_side_effects = false;
		this.statements.addAll(statements);
	}

	@Override
	public String toSourceCode() {

		return
				"fn "
						+ this.methodName
						+ " ("
						+ this.arguments.stream()
						.map(argument -> argument.toSourceCode())
						.collect(Collectors.joining(","))
						+ ") -> "
						+ this.returnType.toSourceCode()
						+ " {"
						+ "\n"
						+ this.statements.stream()
						.map(statement -> "\t" + statement.toSourceCode())
						.collect(Collectors.joining("\n"))
						+ "\n"
						+ "}"
				;
	}
}

File: dragon-parser/src/main/compiler/parsing/astnodes/nonterminal/upperscopes/MethodNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "MethodNode.hpp"

class MethodNode : IASTNode {

public:
	bool isPublic = true;
	bool hasSideEffects;
	TypeNode returnType;
	string methodName;
	vector<DeclaredArgumentNode> arguments;
	vector<StatementNode> statements;

	MethodNode(TokenList tokens, bool debug) throws Exception {
		if (debug) {
			System.out.println("try to parse " + this.getClass().getSimpleName() + " from '" + tokens.toSourceCodeFragment() + "'");
		}

		TokenList copy = tokens.copy();

		copy.expectAndConsumeOtherWiseThrowException(new FnToken());

		this.methodName = new IdentifierNode(copy).identifier;

		copy.expectAndConsumeOtherWiseThrowException(new LParensToken());

		//while there is no ')' up, continue parsing arguments
		IToken next = copy.get(0);
		while (!(next instanceof RParensToken)) {
			if (this.arguments.size() > 0) {
				copy.expectAndConsumeOtherWiseThrowException(new CommaToken());
			}
			this.arguments.add(new DeclaredArgumentNode(copy));
			next = copy.get(0);
		}

		copy.expectAndConsumeOtherWiseThrowException(new RParensToken());

		if (copy.head() instanceof ArrowToken) {
			ArrowToken head = (ArrowToken) copy.head();
			this.hasSideEffects = !head.is_functional;
			copy.consume(1);
		} else {
			throw new Exception("expected arrow here");
		}

		this.returnType = new TypeNode(copy);

		copy.expectAndConsumeOtherWiseThrowException(new LCurlyToken());

		IToken statement_next = copy.get(0);
		while (!(statement_next instanceof RCurlyToken)) {
			this.statements.add(new StatementNode(copy));
			statement_next = copy.get(0);
		}

		copy.expectAndConsumeOtherWiseThrowException(new RCurlyToken());

		tokens.set(copy);
	}

};